### PR TITLE
Client Side Monitoring Support

### PR DESCRIPTION
**Pull request by awood45**
Assuming the presence of a client side monitoring agent, will by default collect client metrics about requests and send them to that agent via UDP.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/master/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

File: gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      def publish(request_metrics)
+        send_datagram(request_metrics.api_call.to_json)
+        request_metrics.api_call_attempts.each do |attempt|
+          send_datagram(attempt.to_json)
+        end
+      end
+
+      def send_datagram(msg)
+        if @agent_port
+          # Should I consider reuse of sockets?
```
Comment (cjyclaire) [inline gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb]: Is this WIP or a discussion question or enhancement to be made later?

File: gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/request_metrics.rb [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+      class ApiCallAttempt
+        attr_reader :service, :api, :client_id, :version, :timestamp, :fqdn,
+          :region, :user_agent, :access_key, :session_token
+        attr_accessor :request_latency, :http_status_code, :aws_exception_msg,
+          :x_amz_request_id, :x_amz_id_2, :x_amzn_request_id
+
+        def initialize(
+          service,
+          api,
```
Comment (cjyclaire) [inline gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/request_metrics.rb]: Lol love to see long list of variable aligned 🎉 Is the concern there trying to make the ApiCall/ApiCallAttempt to be explicit as they are private used classed?

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+Required for publishing client metrics. The port that the client side monitoring
+agent is running on, where client metrics will be published via UDP.
+      DOCS
+        resolve_client_side_monitoring_port(cfg)
+      end
+
+      option(:client_side_monitoring_publisher,
+        default: ClientSideMonitoring::Publisher.new,
+        docstring: <<-DOCS)
+Allows you to provide a custom client-side monitoring publisher class. By default,
```
Comment (cjyclaire) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb]: So by default `client_side_monitoring_publisher` enabled but without a port unless `client_side_monitoring_port `/ENV is specified? Then actually it's not working by pure default?

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+      private
+      def self.resolve_client_side_monitoring_port(cfg)
+        env_source = ENV["AWS_CSM_PORT"]
+        env_source = nil if env_source == ""
+        cfg_source = Aws.shared_config.csm_port(profile: cfg.profile)
+        env_source || cfg_source
+      end
+
+      def self.resolve_client_side_monitoring(cfg)
+        default = true
```
Comment (cjyclaire) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb]: nit pick, the `default` variable is not used?

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            service: service_id,
+            operation: context.operation.name,
+            client_id: context.config.client_side_monitoring_client_id,
+            timestamp: DateTime.now.strftime('%Q').to_i,
+          )
+          context.metadata[:client_metrics] = request_metrics
+          start_time = Time.now
+          begin
+            resp = @handler.call(context)
+          ensure # Anything different in case of failure?
```
Comment (cjyclaire) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb]: Also feel same? as we are tracking those latency data metrics anyway?

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_send_plugin.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+          if headers.include?("x-amzn-request-id")
+            call_attempt.x_amzn_request_id = headers["x-amzn-request-id"]
+          end
+          call_attempt.http_status_code = context.http_response.status_code
+          if e = resp.error
+            e_name = _extract_error_name(e)
+            e_msg = e.message
+            call_attempt.aws_exception_msg = "#{e_name}: #{e_msg}"
+          else
+            # Do I need anything here?
```
Comment (cjyclaire) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_send_plugin.rb]: This this comment something to be removed? Or for discussion?
I image for success requests, at this point, all data need have been collected already?

File: gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
+require 'thread'
+require 'socket'
+
+module Aws
+  module ClientSideMonitoring
+    class Publisher
+      attr_reader :agent_port
+
+      def initialize(opts = {})
+        @agent_port = opts[:agent_port]
```
[the remark this answers is not in the document]
Comment (awood45) [reply gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb]: I think this is a mental miss for me. When I wrote the plugin add-on code, I assumed perhaps we could take on the agent port on the fly, but never wired that up. I think we should require the port at client creation time.

File: gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb (hunk 1 above)
Comment (awood45) [reply gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb]: I need to clean this up - in practice this hasn't been an issue.

File: gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/request_metrics.rb (hunk 2 above)
Comment (awood45) [reply gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/request_metrics.rb]: I should add the `# @api private` tag, agreed.

File: gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+      def send_datagram(msg)
+        if @agent_port
+          # Should I consider reuse of sockets?
+          socket = UDPSocket.new
+          begin
+            socket.connect("127.0.0.1", @agent_port)
+            socket.send(msg, 0)
+          rescue Errno::ECONNREFUSED
+            # Drop on the floor?
```
[the remark this answers is not in the document]
Comment (awood45) [reply gems/aws-sdk-core/lib/aws-sdk-core/client_side_monitoring/publisher.rb]: One issue is that the logging could get noisy. I need to circle back on this offline.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb (hunk 4 above)
Comment (awood45) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb]: Yep that should be cleaned up.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_send_plugin.rb (hunk 6 above)
Comment (awood45) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_send_plugin.rb]: I think so, I basically forgot to clear out some self-discussion comments before sending up the PR. Will clean up for second round.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb (hunk 5 above)
Comment (awood45) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb]: Tracked in a different plugin, regardless of call success. The spec tests should reflect this.

**Pull request merged**